In [1]:
import sys 
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt 
from scipy import stats 
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer 
from sklearn.impute import IterativeImputer 
from sklearn.impute import KNNImputer
from impyute.imputation.cs import mice 
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import IsolationForest




Bad key "text.kerning_factor" on line 4 in
/home/somesh/anaconda3/envs/tf-gpu/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test_patch.mplstyle.
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.1.3/matplotlibrc.template
or from the matplotlib source distribution


ModuleNotFoundError: No module named 'impyute'

In [2]:
#Load DataA.csv 
dataA = pd.read_csv(r'C:\Users\tonkh\OneDrive\Desktop\ECE 657A\Assignment\ece_657a_assignments-master\ece_657a_assignments-master\Assignment_1\Datasets\DataA.csv')

In [3]:
print("\n dataA: \n", dataA)


 dataA: 
        Unnamed: 0  fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7  fea.8  \
0               1 -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0  948.0   
1               2 -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0  939.0   
2               3 -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0  936.0   
3               4 -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0  918.0   
4               5 -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0  939.0   
...           ...    ...    ...    ...    ...     ...    ...    ...    ...   
18995       18996    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18996       18997    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18997       18998    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18998       18999    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   
18999       19000    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN   

       fea.9  ...  fea.72  fea.73  fea.74  fea.75  f

#### As we can see in the above dataset there are many missing values in each of the column. 

In [4]:
# Display all columns
pd.set_option('display.max_columns', None)

## Shape of the Dataset

In [5]:
print("\n Dimension of DataA: ",dataA.shape)


 Dimension of DataA:  (19000, 82)


#### The Dataset has 19000 samples and 82 features. However there is a Unamed column 'Unnamed: 0' which is just the count which needs to be removed from the dataset 

## Checking all the Columns

In [6]:
# Columns of DataA
print(" \n Columns of DataA: \n",dataA.columns)


 Columns of DataA: 
 Index(['Unnamed: 0', 'fea.1', 'fea.2', 'fea.3', 'fea.4', 'fea.5', 'fea.6',
       'fea.7', 'fea.8', 'fea.9', 'fea.10', 'fea.11', 'fea.12', 'fea.13',
       'fea.14', 'fea.15', 'fea.16', 'fea.17', 'fea.18', 'fea.19', 'fea.20',
       'fea.21', 'fea.22', 'fea.23', 'fea.24', 'fea.25', 'fea.26', 'fea.27',
       'fea.28', 'fea.29', 'fea.30', 'fea.31', 'fea.32', 'fea.33', 'fea.34',
       'fea.35', 'fea.36', 'fea.37', 'fea.38', 'fea.39', 'fea.40', 'fea.41',
       'fea.42', 'fea.43', 'fea.44', 'fea.45', 'fea.46', 'fea.47', 'fea.48',
       'fea.49', 'fea.50', 'fea.51', 'fea.52', 'fea.53', 'fea.54', 'fea.55',
       'fea.56', 'fea.57', 'fea.58', 'fea.59', 'fea.60', 'fea.61', 'fea.62',
       'fea.63', 'fea.64', 'fea.65', 'fea.66', 'fea.67', 'fea.68', 'fea.69',
       'fea.70', 'fea.71', 'fea.72', 'fea.73', 'fea.74', 'fea.75', 'fea.76',
       'fea.77', 'fea.78', 'fea.79', 'fea.80', 'fea.81'],
      dtype='object')


#### Removing 'Unnamed: 0' columns from the dataset  

In [7]:
dataA_updated = dataA.drop(['Unnamed: 0'], axis=1)
print("Updated Dataset after removal of 'Unnamed: 0' column: \n", dataA_updated.head(20))

Updated Dataset after removal of 'Unnamed: 0' column: 
    fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7   fea.8  fea.9  fea.10  \
0  -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0   948.0  333.0   -19.0   
1  -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0   939.0  316.0     9.0   
2  -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0   936.0  315.0   -10.0   
3  -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0   918.0  338.0    34.0   
4  -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0   939.0  350.0    60.0   
5  -171.0  432.0  924.0 -179.0  1011.0   85.0 -204.0   945.0  336.0    94.0   
6  -169.0  429.0  949.0 -175.0  1007.0  102.0 -188.0   914.0  322.0   154.0   
7  -160.0  423.0  927.0 -195.0   996.0  123.0 -213.0   925.0  302.0   128.0   
8  -163.0  432.0  929.0 -178.0   994.0  101.0 -186.0   946.0  296.0   166.0   
9  -156.0  415.0  936.0 -186.0  1014.0  111.0 -195.0   960.0  280.0   202.0   
10 -153.0  413.0  923.0 -187.0   993.0   91.0 -193.0   970.0  282.0   233.0 

In [8]:
# Dimension of the updated Dataset
print("\n Dimension of Updated DataA: ",dataA_updated.shape)


 Dimension of Updated DataA:  (19000, 81)


#### The Column 'Unnamed: 0' has been removed from the dataset as we can see in the above data and now the dimension of the dataset has become (19000, 81) which means it has 19000 samples and 81 features in total

## Checking Missing Values in the Dataset 


In [9]:
# Missing Values in the Dataset
print("\n Missing values in the DataA: \n",dataA_updated.isnull())


 Missing values in the DataA: 
       fea.1  fea.2  fea.3  fea.4  fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0      False  False  False  False  False  False  False  False  False   False   
1      False  False  False  False  False  False  False  False  False   False   
2      False  False  False  False  False  False  False  False  False   False   
3      False  False  False  False  False  False  False  False  False   False   
4      False   True  False  False  False  False  False  False  False   False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   
18995   True   True   True   True   True   True   True   True   True    True   
18996   True   True   True   True   True   True   True   True   True    True   
18997   True   True   True   True   True   True   True   True   True    True   
18998   True   True   True   True   True   True   True   True   True    True   
18999   True   True   True   True   True   True   True   True   True    True   

       

#### Here We can see in the dataset there are many missing values as the dataset contains value as 'True' in some rows and columns after applying 'dataA_updated.isnull()'. 

## Number of Missing values per columns

In [10]:
# Find the Total number of missing values in each column
print("\n Number of missing values in each Column: \n", dataA_updated.isnull().sum())


 Number of missing values in each Column: 
 fea.1     1187
fea.2     1188
fea.3     1187
fea.4      800
fea.5      800
          ... 
fea.77     773
fea.78     773
fea.79     773
fea.80     773
fea.81     773
Length: 81, dtype: int64


#### As we can see from the above table that the dataset contains too many missing values and each feature has different number of missing values. As from the table it is clear that feature.1 has 1187 missing values, feature.2 has 1188 missing values while from feature.37 to feature.81 all have a missing value of 773 which is the lowest value missing among all the features. Whereas feature.34 to feature.36 have 18999 missing values which are highest among all the features in the entire dataset. 

## Total number of Missing values in the dataset

In [11]:
print("\n Total number of Missing values in the dataset: ", dataA_updated.isnull().sum().sum())


 Total number of Missing values in the dataset:  124053


#### In total there are 124053 missing values in the dataset

In [12]:
# Heat Map showing number of Missing Values per Column
plt.subplots(figsize=(20,10))
sns.heatmap(dataA_updated.isnull())
plt.show()

#### As it can be seen from the heat map that Feature 34, 35 and 36 are missing so these features have to be removed

## Statistical Description of the Dataset with missing values

In [13]:
# Statistical analysis of the Dataset
dataA_updated.describe()

fea.1         fea.2         fea.3         fea.4         fea.5  \
count  17813.000000  17812.000000  17813.000000  18200.000000  18200.000000   
mean    -132.812384    698.264485    597.541402   -307.128462    909.548077   
std      284.183187    375.672475    396.654659    183.151634    193.963300   
min    -2724.000000   -855.000000  -2196.000000  -1365.000000   -245.000000   
25%     -179.000000    360.000000    304.000000   -409.000000    860.000000   
50%     -100.000000    811.000000    574.000000   -266.000000    969.500000   
75%      -15.000000    984.000000    955.000000   -167.000000   1006.000000   
max     1887.000000   2531.000000   2941.000000    609.000000   1833.000000   

              fea.6         fea.7         fea.8         fea.9        fea.10  \
count  18200.000000  18099.000000  18099.000000  18099.000000  18043.000000   
mean     -32.760824     61.974363    899.313498     81.650478    356.638752   
std      254.001018    317.393784    196.829252    327.904371    343.131382   
min     -920.000000  -1580.000000   -149.000000  -1624.000000  -1792.000000   
25%     -144.000000   -131.000000    854.000000   -155.000000    158.000000   
50%      -39.000000     70.000000    946.000000     41.000000    380.000000   
75%       45.000000    251.000000    997.000000    315.000000    583.000000   
max     1215.000000   1490.000000   1682.000000   1096.000000   2202.000000   

             fea.11        fea.12        fea.13        fea.14        fea.15  \
count  18044.000000  18044.000000  17950.000000  17950.000000  17950.000000   
mean     330.971569    584.833130    583.589861    421.624290    -38.719833   
std      543.517364    282.078053    307.610433    543.293018    495.913749   
min    -1545.000000  -1079.000000  -1710.000000  -1120.000000  -1634.000000   
25%     -142.000000    370.000000    347.000000   -131.000000   -484.750000   
50%      395.000000    610.000000    628.000000    524.000000    -14.000000   
75%      786.000000    814.000000    818.000000    886.000000    274.000000   
max     2047.000000   2152.000000   2408.000000   2212.000000   2620.000000   

             fea.16        fea.17        fea.18        fea.19        fea.20  \
count  18202.000000  18202.000000  18202.000000  17964.000000  17964.000000   
mean      71.807768    780.358257   -461.605043    274.476843    893.086451   
std      258.937246    291.642323    261.243774    298.191766    216.179301   
min    -1089.000000   -289.000000  -1379.000000  -2269.000000   -786.000000   
25%        6.000000    696.000000   -649.000000    252.000000    878.000000   
50%       94.000000    905.000000   -398.000000    302.000000    921.000000   
75%      180.000000    967.000000   -282.000000    356.000000    945.000000   
max      878.000000   1653.000000   1292.000000   2335.000000   2466.000000   

             fea.21        fea.22        fea.23        fea.24        fea.25  \
count  17964.000000  17677.000000  17677.000000  17677.000000  18106.000000   
mean     -89.299933   -622.139334    233.764044   -131.400068   -360.661825   
std      221.071614    359.196753    587.480524    460.080469    223.329186   
min    -1944.000000  -3172.000000  -2324.000000  -2734.000000  -1873.000000   
25%     -192.000000   -876.000000   -193.000000   -468.000000   -475.000000   
50%      -70.000000   -655.000000    267.000000   -136.000000   -342.000000   
75%       31.000000   -371.000000    731.000000    184.000000   -205.000000   
max     1968.000000   2229.000000   2677.000000   2401.000000    711.000000   

             fea.26        fea.27        fea.28        fea.29        fea.30  \
count  18106.000000  18106.000000  18106.000000  18106.000000  18106.000000   
mean     875.943278    269.614934   -570.500166    432.334420    427.547277   
std      253.761954    371.407492    307.329729    497.493607    204.477583   
min     -442.000000  -1430.000000  -1933.000000  -1103.000000   -765.000000   
25%      770.000000     11.000000   -825.000000    -59.0000

#### As we can see from the above statistical description of the dataset it clear that the rows of feature.34, feature.35 and feature.36 is entirely missing so there is no statistical value for those columns. We have to remove the entire feature.34, feature.35 and feature.36 from the dataset.

## Drop feature.34, feature.35 and feature.36

In [14]:
dataA_updated_redDim = dataA_updated.drop(['fea.34', 'fea.35', 'fea.36'], axis=1)
print("\n Dimension of Reduced Dataset: ", dataA_updated_redDim.shape)
print("\n Reduced Dataset: \n", dataA_updated_redDim)


 Dimension of Reduced Dataset:  (19000, 78)

 Reduced Dataset: 
        fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0     -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0  948.0  333.0   -19.0   
1     -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0  939.0  316.0     9.0   
2     -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0  936.0  315.0   -10.0   
3     -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0  918.0  338.0    34.0   
4     -171.0    NaN  929.0 -189.0  1004.0  104.0 -198.0  939.0  350.0    60.0   
...      ...    ...    ...    ...     ...    ...    ...    ...    ...     ...   
18995    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18996    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18997    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18998    NaN    NaN    NaN    NaN     NaN    NaN    NaN    NaN    NaN     NaN   
18999    NaN    NaN    NaN    NaN     NaN  

#### As it can be seen from the above dataset that the Feature.34, Feature.35 and feature.36 has been dropped from the dataset.

In [15]:
# Total missing value check after removing the Features
print("\n Number of missing values after feature removal: ", dataA_updated_redDim.isnull().sum().sum())


 Number of missing values after feature removal:  67056


#### The dataset still contains 67056 missing values in the other Feature sets. So we will apply Imputation of missing values learned during lectures. 

# Imputing Missing Values
### 1.A Replace missing data by Mean values for each columns

In [16]:
# Using Simple Imputation Technique for replacing the missing values with mean values 
si = SimpleImputer(strategy='mean')
si.fit(dataA_updated_redDim)
dataA_Mean_SiImputation = si.transform(dataA_updated_redDim)


In [17]:
print("\n Data after Mean Imputation: \n", dataA_Mean_SiImputation)


 Data after Mean Imputation: 
 [[-153.          414.          939.         ...  -29.
    36.           24.        ]
 [-150.          420.          939.         ...  -31.
    47.            3.        ]
 [-160.          432.          941.         ...  -38.
    20.            0.        ]
 ...
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]]


### 1.B Using fillna for mean imputation and Checking for Missing Values after Imputation

In [18]:
# Copy data in another variable
data_forMeanImputing = dataA_updated_redDim.copy(deep = True)

In [19]:
# Mean Method for fill na 
data_forMeanImputing.fillna(data_forMeanImputing.mean(), inplace=True)
print("\n Data after Mean imputation: \n",data_forMeanImputing)


 Data after Mean imputation: 
             fea.1       fea.2       fea.3       fea.4        fea.5  \
0     -153.000000  414.000000  939.000000 -161.000000  1007.000000   
1     -150.000000  420.000000  939.000000 -177.000000  1008.000000   
2     -160.000000  432.000000  941.000000 -162.000000   982.000000   
3     -171.000000  432.000000  911.000000 -174.000000   999.000000   
4     -171.000000  698.264485  929.000000 -189.000000  1004.000000   
...           ...         ...         ...         ...          ...   
18995 -132.812384  698.264485  597.541402 -307.128462   909.548077   
18996 -132.812384  698.264485  597.541402 -307.128462   909.548077   
18997 -132.812384  698.264485  597.541402 -307.128462   909.548077   
18998 -132.812384  698.264485  597.541402 -307.128462   909.548077   
18999 -132.812384  698.264485  597.541402 -307.128462   909.548077   

            fea.6       fea.7       fea.8       fea.9      fea.10      fea.11  \
0       99.000000 -210.000000  948.000000  333

In [20]:
# Check for missing values in the dataset
print("\n Missing Value check in the dataset: ", data_forMeanImputing.isnull().sum().sum())


 Missing Value check in the dataset:  0


In [21]:
# Statistical Table for the Dataset after Mean imputation
print(data_forMeanImputing.describe())

fea.1         fea.2         fea.3         fea.4         fea.5  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean    -132.812384    698.264485    597.541402   -307.128462    909.548077   
std      275.162567    363.737566    384.063939    179.254138    189.835729   
min    -2724.000000   -855.000000  -2196.000000  -1365.000000   -245.000000   
25%     -172.000000    366.000000    322.750000   -402.000000    868.000000   
50%     -110.000000    756.000000    597.541402   -277.000000    964.000000   
75%      -20.000000    977.000000    950.000000   -169.000000   1005.000000   
max     1887.000000   2531.000000   2941.000000    609.000000   1833.000000   

              fea.6         fea.7         fea.8         fea.9        fea.10  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     -32.760824     61.974363    899.313498     81.650478    356.638752   
std      248.595835    309.776406    192.105395    320.034742    

#### As we can see from the above dataset that the missing values have been filled with mean values after applying simple Imputation  

#### It is clear that all the missing values from the dataset are filled with mean values for each column. However there are merits and dmerits of using Imputaion with Mean values they are shown as: 

Merits: 
------
1. It is easy and faster to implement.  
2. It works well with small dataset.

Demerits: 
---------
1. It doesn't factor the correlation between the features.
2. It doesn't encounter the Uncertainity in the imputation
3. It is not very accurate.
4. It is susceptible to Outliers



## 2. Median Imputation

In [22]:
# Copy Data for Median Imputation
dataA_updatedForMedian = dataA_updated_redDim.copy(deep=True)

In [23]:
si_mean = SimpleImputer(strategy='median')
si_mean.fit(dataA_updatedForMedian)
dataA_SiMedianImputation = si_mean.transform(dataA_updated_redDim)

In [24]:
print("\n Data after Median Imputation: \n", dataA_SiMedianImputation)



 Data after Median Imputation: 
 [[-153.  414.  939. ...  -29.   36.   24.]
 [-150.  420.  939. ...  -31.   47.    3.]
 [-160.  432.  941. ...  -38.   20.    0.]
 ...
 [-100.  811.  574. ...  -29.    4.   19.]
 [-100.  811.  574. ...  -29.    4.   19.]
 [-100.  811.  574. ...  -29.    4.   19.]]


Merits:
-------
#### The Median imputation is not susceptible to Outliers so it is a better way for finding missing values

Demerits: 
---------
#### It doesn't factor the correlation between the features. 


## 3. Imputation using KNN

In [25]:
knnimpute = KNNImputer(n_neighbors=2)
dataA_knnImputation = knnimpute.fit_transform(dataA_updated_redDim)

In [26]:
print("\n Data after KNN imputation for missing values: \n ", dataA_knnImputation)


 Data after KNN imputation for missing values: 
  [[-153.          414.          939.         ...  -29.
    36.           24.        ]
 [-150.          420.          939.         ...  -31.
    47.            3.        ]
 [-160.          432.          941.         ...  -38.
    20.            0.        ]
 ...
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]
 [-132.81238421  698.26448462  597.54140235 ...  -18.09952269
     4.67125693   20.72683382]]


#### The Imputed Values obatined from KNN based method is very similar to the Imputed values obtained by using Mean Imputation.

Merits
------
1. much more accurate than Mean/Median

Demerits
--------
1. It is sensitive to the outliers
2. It is computaionally very Expensive 


## 4. Imputation using Linear Interpolation

In [27]:
# Copy the data for Linear Interpolation 
data_linInterpolate = dataA_updated_redDim.copy(deep=True)

In [28]:

dataA_interpolateImpute = data_linInterpolate.interpolate(method='linear', axis=0)
print("\n Data after Linear Interpolation: \n",dataA_interpolateImpute)


 Data after Linear Interpolation: 
        fea.1  fea.2  fea.3  fea.4   fea.5  fea.6  fea.7  fea.8  fea.9  fea.10  \
0     -153.0  414.0  939.0 -161.0  1007.0   99.0 -210.0  948.0  333.0   -19.0   
1     -150.0  420.0  939.0 -177.0  1008.0  103.0 -207.0  939.0  316.0     9.0   
2     -160.0  432.0  941.0 -162.0   982.0   98.0 -198.0  936.0  315.0   -10.0   
3     -171.0  432.0  911.0 -174.0   999.0  115.0 -187.0  918.0  338.0    34.0   
4     -171.0  432.0  929.0 -189.0  1004.0  104.0 -198.0  939.0  350.0    60.0   
...      ...    ...    ...    ...     ...    ...    ...    ...    ...     ...   
18995    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0     0.0   
18996    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0     0.0   
18997    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0     0.0   
18998    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0     0.0   
18999    0.0    0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0   

In [29]:
# Check for missing values 
print("\n Missing values after Linear Interpolation: ",dataA_interpolateImpute.isnull().sum().sum())


 Missing values after Linear Interpolation:  0


In [30]:
# Statistical Analysis for the Dataset after Linear Interpolation
print(dataA_interpolateImpute.describe())

fea.1         fea.2         fea.3         fea.4         fea.5  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean    -124.740553    679.817605    563.032474   -294.196737    871.251316   
std      279.158229    401.075991    408.463091    189.569822    263.449167   
min    -2724.000000   -855.000000  -2196.000000  -1365.000000   -245.000000   
25%     -173.000000    334.000000    250.000000   -402.000000    836.000000   
50%      -92.000000    794.000000    536.000000   -255.000000    964.000000   
75%        0.000000    983.000000    950.000000   -156.000000   1005.000000   
max     1887.000000   2531.000000   2941.000000    609.000000   1833.000000   

              fea.6         fea.7         fea.8         fea.9        fea.10  \
count  19000.000000  19000.000000  19000.000000  19000.000000  19000.000000   
mean     -31.381421     56.161263    859.519211     81.482737    341.387000   
std      248.682889    313.189606    264.930665    323.747494    

#### From the Statical Analysis obtained after Imputing with Linear Iterpolation we observe that the mean and 2 standard away mean(50%) for Linear Interpolation method values are different from that obtained after Mean/Median Imputation on the dataset.

## 5. Multivariate Imputation using Iterative Imputation 

In [31]:
# Initialize the IterativeImputation
ii = IterativeImputer(initial_strategy = 'median', random_state=42, imputation_order = 'descending', add_indicator = True)
ii.fit(dataA_updated_redDim)
dataA_IterImpute = ii.transform(dataA_updated_redDim)


In [32]:
print("\n Data after Iterative Imputation: \n", dataA_IterImpute)


 Data after Iterative Imputation: 
 [[-153.          414.          939.         ...    0.
     0.            0.        ]
 [-150.          420.          939.         ...    0.
     0.            0.        ]
 [-160.          432.          941.         ...    0.
     0.            0.        ]
 ...
 [-100.08514228  729.70155501  605.13668153 ...    1.
     1.            1.        ]
 [-100.08514228  729.70155501  605.13668153 ...    1.
     1.            1.        ]
 [-100.08514228  729.70155501  605.13668153 ...    1.
     1.            1.        ]]


In [33]:
# Number of Features with missing values 
print("\n Features with missing values before Imputation: ",ii.n_features_with_missing_)


 Features with missing values before Imputation:  78


#### It is evident from the output obtained after Multivariate Imputation using Iterative Imputation that the values are distinct from what we have obtained from Mean, Median, KNN and Linear Interpolation. 

## 6. [MICE](http://scholar.google.ca/scholar_url?url=https://www.jstatsoft.org/article/view/v045i04/v45i04.pdf&hl=en&sa=X&scisig=AAGBfm2i0J4zMPMI6FJgalzvk2DxTDu65g&nossl=1&oi=scholarr)

#### It works by filling missing data multiple times. As given in the [paper](http://scholar.google.ca/scholar_url?url=https://www.jstatsoft.org/article/view/v045i04/v45i04.pdf&hl=en&sa=X&scisig=AAGBfm2i0J4zMPMI6FJgalzvk2DxTDu65g&nossl=1&oi=scholarr) Multiple Imputation is much better than that of Single Imputation as it improves the measure of Uncertainity of the missing values. 

#### In the library of impyute the algorithm of [MICE] uses LinearRegression for Convergence of missing values for the Dataset.

In [34]:
dataA_MICEImputation = mice(dataA_updated_redDim.values)

In [35]:
print("\n Data after MICE Imputation: \n", dataA_MICEImputation)

 ### Output from MICE Algorithm

 Data after MICE Imputation: 
 [[-153.          414.          939.         ...  -29.
    36.           24.        ]
 [-150.          420.          939.         ...  -31.
    47.            3.        ]
 [-160.          432.          941.         ...  -38.
    20.            0.        ]
 ...
 [-130.36492054  705.07225938  589.2263699  ...  -18.09952269
     4.67125693   20.72683382]
 [-130.36492054  705.07225938  589.2263699  ...  -18.09952269
     4.67125693   20.72683382]
 [-130.36492054  705.07225938  589.2263699  ...  -18.09952269
     4.67125693   20.72683382]]

#### The Output obtained from the above algorithm differs from that of all the above methods and since the basis of this algorithm uses LinearRegression it should have more accurate data for a given value. 

In [36]:
# Check for Missing values in the dataset using Heatmap
plt.subplots(figsize=(20,10))
sns.heatmap(dataA_MICEImputation.isnull())
plt.show()


#### As it can be seen from the heat map that there are no missing values in the dataset.

# Outlier Detection in the Dataset

In [37]:
# Total Outlier count in the Dataset 
dataA_std = dataA_MICEImputation.std()
dataA_mean = dataA_MICEImputation.mean()

dataA_data_outliers=dataA_MICEImputation.apply(lambda y: np.abs(y-dataA_mean)>(3*dataA_std),axis=1)
print("Total number of ouliers in each feature are: \n",dataA_data_outliers.sum())

NameError: name 'dataA_MICEImputation' is not defined

In [38]:
# Total Number of outlier in the dataset 
print("Total number of ouliers in the dataset: ",dataA_data_outliers.sum().sum())

NameError: name 'dataA_data_outliers' is not defined

#### So as seen from the above analysis that there are exactly 16323 outlier in the dataset. In addition, a count for number of outlier in each feature is shown.

### 1. Using Z-Score for Outlier Detection

In [39]:
# Detect Outlier using ZScore
outlier_value = np.abs(stats.zscore(dataA_MICEImputation))
print("Outlier: \n",outlier_value)

NameError: name 'dataA_MICEImputation' is not defined

In [40]:
standard_thershold = 3 
print("\n Position of Outliers in the Dataset: \n",np.where(outlier_value>standard_thershold))

NameError: name 'outlier_value' is not defined

In [41]:
print("\n Outlier at row 60 and column 1: ",outlier_value[60][1])

NameError: name 'outlier_value' is not defined

#### The above says that there is an outlier at row: 60 in column: 1.

### 2. IQR Method for Outlier detection 

In [42]:
# find the Quartiles for the dataset 
Q1 = dataA_MICEImputation.quantile(0.25)
Q3 = dataA_MICEImputation.quantile(0.75)
IQR = Q3 - Q1
print("The InterQuartile Range for Each Features is: \n ", IQR)

NameError: name 'dataA_MICEImputation' is not defined

In [43]:
# Detect the Oultiers using IQR technique
print("\n Outliers in the complete dataset: \n", (dataA_MICEImputation < (Q1 - 1.5 * IQR))|(dataA_MICEImputation > (Q3 + 1.5 * IQR)))

NameError: name 'dataA_MICEImputation' is not defined

#### The value where there is a True label indicated the presence of an Outlier and False when there are no Outliers.

### 3. Outlier detection using IsolationForest

In [44]:
# IsolationForest Initialization 
isoforest = IsolationForest(max_samples="auto", n_jobs=-1)
dataA_outliers_isoforest = isoforest.fit_predict(dataA_MICEImputation)

NameError: name 'IsolationForest' is not defined

In [45]:
print("\n Outliers from IsolationForest: \n", dataA_outliers_isoforest)

NameError: name 'dataA_outliers_isoforest' is not defined

In [46]:
print("Shape of IsolationForest Dataset: ",dataA_outliers_isoforest.shape)

NameError: name 'dataA_outliers_isoforest' is not defined

In [47]:
# Display all the labels from IsolationForest
count = 0
for value in range(dataA_outliers_isoforest.shape[0]):
    if dataA_outliers_isoforest[value] == -1:
        sys.stderr.write("\n Outliers for each label: {0}".format(dataA_outliers_isoforest[value]))
        sys.stderr.flush()
        count += 1

print("\n Total number Outliers detected by IsolationForest is: ", count)

NameError: name 'dataA_outliers_isoforest' is not defined

#### The label given by Isolation Forest method gives us whether a Outlier is present or not. If the label is -1 then there is a outlier whereas when the label is +1 then there are no outlier or it is an inlier. As seen from the above analysis that IsolationForest gives us the total number of Outliers in the dataset which is 1158 and it is far less than Z-score and IQR method. 

## Removing Outliers from the Imputed Dataset

In [48]:
# Remove outliers in the MICE Obtained Dataset
dataA_OutlierRemovedMice = dataA_MICEImputation[(z < 3).all(axis=1)]

NameError: name 'dataA_MICEImputation' is not defined

#### In the above code the outliers have been removed from MICE Imputed dataset.

# Feature.9 and Feature.24 plot
### Before Normalization

In [50]:
# Plot of Feature.9 with MICE imputed dataset
n, bins, patches = plt.hist(x=dataA_OutlierRemovedMice[:, 8], bins='auto', color='m', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.title('Feature.9 Plot before Normalization')
plt.text(-1500,1200, r'$\mu=81.650 , \sigma= 320.035$')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'dataA_OutlierRemovedMice' is not defined

In [51]:
# Plot of Feature.24 with MICE imputed dataset
n, bins, patches = plt.hist(x=dataA_OutlierRemovedMice[:, 23], bins='auto', color='r', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.text(-2500,1750, r'$\mu=-131.400, \sigma= 443.773$')
plt.title('Feature.24 Plot before Normalization')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'dataA_OutlierRemovedMice' is not defined

# Data Normalization
## 1. Max-Min Normalization

In [52]:
MMscaler = MinMaxScaler()
MMscaler.fit(dataA_OutlierRemovedMice)
data_normalizedMM = MMscaler.transform(dataA_OutlierRemovedMice)

NameError: name 'MinMaxScaler' is not defined

In [53]:
print("\n Max-Min Normalized dataset: \n", data_normalizedMM)

NameError: name 'data_normalizedMM' is not defined

In [54]:
print("Data of Feature.9: \n",data_normalizedMM[:,9])

NameError: name 'data_normalizedMM' is not defined

In [55]:
# Plot of Feature.9 with MICE imputed dataset
n, bins, patches = plt.hist(x=data_normalizedMM[:, 8], bins='auto', color='m', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.title('Feature.9 Plot after Min-Max Normalization')
plt.text(0,1200, r'$\mu=81.650 , \sigma= 320.035$')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'data_normalizedMM' is not defined

In [56]:
# Plot of Feature.24 with MICE imputed dataset
n, bins, patches = plt.hist(x=data_normalizedMM[:, 23], bins='auto', color='r', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.title('Feature.24 Plot after Min-Max Normalization')
plt.text(0,1200, r'$\mu=81.650 , \sigma= 320.035$')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'data_normalizedMM' is not defined

## 2. Z-Score Normalization (Standardization)

In [57]:
# Z-Score Normalization
ss = StandardScaler()
ss.fit(dataA_OutlierRemovedMice)
dataA_standardised = ss.transform(dataA_OutlierRemovedMice)

NameError: name 'StandardScaler' is not defined

In [58]:
print("Standardized Dataset: \n", dataA_standardised)

NameError: name 'dataA_standardised' is not defined

In [59]:
# Plot of Feature.9 with MICE imputed dataset
n, bins, patches = plt.hist(x=dataA_standardised[:, 8], bins='auto', color='y', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.title('Feature.9 Plot after Z-Score Normalization')
plt.text(-2,1200, r'$\mu=81.650 , \sigma= 320.035$')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'dataA_standardised' is not defined

In [60]:
# Plot of Feature.24 with MICE imputed dataset
n, bins, patches = plt.hist(x=dataA_standardised[:, 23], bins='auto', color='k', alpha=0.7, rwidth=0.85)
plt.grid(axis='y', alpha=0.75)
plt.title('Feature.24 Plot after Z-Score Normalization')
plt.text(-6,1200, r'$\mu=81.650 , \sigma= 320.035$')
maxfreq = n.max()
plt.ylim(ymax=np.ceil(maxfreq/10) * 10 if maxfreq % 10 else maxfreq + 10)

NameError: name 'dataA_standardised' is not defined

## Analysis: 
#### From the plot generated before max min and z-score normalization the x-axis for the plot by feature.9 ranges from -1500 to 1000 and that of the feature.24 ranges from -2000 to 2000. After max-min normalization the range of the x-axis changes and it becomes normalized between 0 and 1 for both the plot for feature.9 and feature.24. Where the plot generated after Z-score normalization the x-range ranges from -4 to 4 for both the features. However there is no change in the y label of the plots before and after normalization for both max-min and z-score normalization.